python/hidet/graph/ops/definitions/normalize/norm_f16.py

```diff
@@ -0,0 +1,272 @@
+from hidet.lang import data_type, TensorType, i32, f16, attrs, tensor
+from hidet.lang.cuda import blockIdx, threadIdx, register_tensor, syncthreads
+from hidet.graph.ops.definitions.utils import Operator, Tensor, normalize_dim
+from hidet.graph.ops.definitions.utils import input_like
+from hidet.utils import prod
+from hidet.ir import primitives as prim
+from .norm import NormalizeTask
+
+
+class NormalizeF16Task(NormalizeTask):
+    """
+    Performs the following operation in float16 precision
+        mean = x.mean(dims, keep_dim=True)
+        x = x - mean
+        variance = square(x).mean(dims, keep_dim=True)
+        x = x * rsqrt(variance + epsilon)
+    """
+
+    def implement_cuda(self, working_dir: str) -> IRModule:
+        import hidet
+        import math
+
+        x, y = self.inputs[0], self.outputs[0]
+        input_shape: List[int] = list(x.const_shape)
+        dims = self.dims
+
+        spatial_shape = [v for i, v in enumerate(input_shape) if i not in dims]
+        reduce_shape = [input_shape[i] for i in dims]
+        dim_zeros = [0] * len(dims)
+
+        reduce_extent = prod(reduce_shape)
+
+        warp_size = 64  # coleased loads
+        block_size = min(max(warp_size, reduce_extent), 1024)
+        repeat_reduction = math.ceil(reduce_extent / block_size)
+
+        task_layout = spatial(*spatial_shape)
+        grid_size = task_layout.num_workers
+
+        accumulate_dtype = data_type(self.attrs['accumulate_dtype'])
+
+        shm_count = math.ceil(block_size / warp_size)
+        used_smem_bytes_per_block = shm_count
+
+        stages = math.ceil(math.log(block_size) / math.log(warp_size))
+        assert stages <= 2
+
+        with hidet.script_module() as module:
+
+            @hidet.script
+            def welford_combine(
+                mean_a: TensorType(dtype=accumulate_dtype, shape=[1]),
+                m2_a: TensorType(dtype=accumulate_dtype, shape=[1]),
+                count_a: TensorType(dtype=i32, shape=[1]),
+                mean_b: TensorType(dtype=accumulate_dtype, shape=[1]),
+                m2_b: TensorType(dtype=accumulate_dtype, shape=[1]),
+                count_b: TensorType(dtype=i32, shape=[1]),
+            ):
+                count = count_a[0] + count_b[0]
+                if count == 0:
+                    return
+                delta = mean_b[0] - mean_a[0]
+
+                mean_a[0] = mean_a[0] + delta * cast(count_b[0], accumulate_dtype) / cast(count, accumulate_dtype)
+                m2_a[0] = (
+                    m2_a[0]
+                    + m2_b[0]
+                    + delta
+                    * delta
+                    * cast(count_a[0], accumulate_dtype)
+                    * cast(count_b[0], accumulate_dtype)
+                    / cast(count, accumulate_dtype)
+                )
+                count_a[0] = count
+
+            @hidet.script
+            def norm_kernel(x: f16[x.const_shape], y: f16[y.const_shape]):
```

Comment:
We might want to read float16x2 at a time in the future instead of single float16, the former one would have better efficiency.

Reply:
you are right, this approach currently only loads 64 bytes coalesced.  I will add it